File: 백준/Gold/1167. 트리의 지름/트리의 지름.py
import sys

input = sys.stdin.readline

def dfs(node, distance):
    global max_distance, farthest_node
    visited[node] = True

    if distance > max_distance:
        max_distance = distance
        farthest_node = node

    for v, w in graph[node]:
        if not visited[v]:
            dfs(v, distance + w)

# 입력값 받기
n = int(input())
graph = [[] for _ in range(n + 1)]
visited = [False] * (n + 1)

for _ in range(n):
    info = list(map(int, input().split()))
    node = info[0]
    for i in range(1, len(info) - 1, 2):
        v, w = info[i], info[i + 1]
        graph[node].append((v, w))

# 첫 번째 DFS를 통해 아무 노드에서 시작해 가장 먼 노드를 찾음
max_distance = 0
farthest_node = 0
dfs(1, 0)

# 두 번째 DFS를 통해 가장 먼 노드로부터 다시 가장 먼 노드를 찾음
# 이것이 트리의 지름이 됨
max_distance = 0
visited = [False] * (n + 1)
dfs(farthest_node, 0)

print(max_distance)

File: 백준/Gold/1167. 트리의 지름/트리의 지름.js
const fs = require('fs');
const input = fs.readFileSync('/dev/stdin').toString().trim().split('\n');

const V = parseInt(input[0], 10);
const graph = Array.from({ length: V + 1 }, () => []);

for (let i = 1; i <= V; i++) {
  const line = input[i].split(' ').map(Number);
  const u = line[0];
  for (let j = 1; j < line.length - 1; j += 2) {
    const v = line[j];
    const w = line[j + 1];
    graph[u].push([v, w]);
  }
}

let maxDistance = 0;
let farthestNode = 0;

// DFS 함수 정의
function dfs(node, distance, visited) {
  visited[node] = true;

  if (distance > maxDistance) {
    maxDistance = distance;
    farthestNode = node;
  }

  for (const [nextNode, weight] of graph[node]) {
    if (!visited[nextNode]) {
      dfs(nextNode, distance + weight, visited);
    }
  }
}

// 1. 임의의 노드(1번)에서 가장 먼 노드를 찾음
let visited = Array(V + 1).fill(false);
dfs(1, 0, visited);

// 2. 그 노드로부터 가장 먼 노드를 찾음
maxDistance = 0;
visited = Array(V + 1).fill(false);
dfs(farthestNode, 0, visited);

console.log(maxDistance);